lib/Adapter/Tolerant/TolerantIndexBuilder.php

```diff
@@ -0,0 +1,155 @@
+        $this->indexNode($info, $node);
+    }
+
+    public function done(): void
+    {
+        $this->index->updateTimestamp();
+    }
+
+    private function indexNode(SplFileInfo $info, Node $node): void
+    {
+        if ($node instanceof ClassDeclaration) {
+            $this->indexClassDeclaration($info, $node);
+            return;
+        }
+
+        if ($node instanceof FunctionDeclaration) {
+            $this->indexFunction($info, $node);
+            return;
+        }
+
+        foreach ($node->getChildNodes() as $childNode) {
+            $this->indexNode($info, $childNode);
+        }
+    }
+
+    private function indexClassDeclaration(SplFileInfo $info, ClassDeclaration $node): void
+    {
+        $record = $this->index->get(ClassRecord::fromName($node->getNamespacedName()->getFullyQualifiedNameText()));
+        assert($record instanceof ClassRecord);
+        $record->withLastModified($info->getCTime());
+        $record->withStart(ByteOffset::fromInt($node->getStart()));
+        $record->withType('class');
+        $record->withFilePath($info->getPathname());
+
+        // de-reference this class
+        foreach ($record->implementedClasses() as $implementedClass) {
+            $implementedRecord = $this->index->get(ClassRecord::fromName($implementedClass));
+
+            if (false === $implementedRecord->removeImplementation($record->fqn())) {
+                continue;
+            }
+
+            $this->index->write($implementedRecord);
+        }
+
+        $record->clearImplements();
+
+        $this->indexClassInterfaces($record, $node);
+        $this->indexBaseClass($record, $node);
+
+        $this->index->write($record);
+    }
+
+    private function indexClassInterfaces(ClassRecord $classRecord, ClassDeclaration $node): void
+    {
+        // @phpstan-ignore-next-line
+        if (null === $interfaceClause = $node->classInterfaceClause) {
+            return;
+        }
+
+        if (null == $interfaceList = $interfaceClause->interfaceNameList) {
+            return;
+        }
+
+        foreach ($interfaceList->children as $interfaceName) {
+            if (!$interfaceName instanceof QualifiedName) {
+                continue;
+            }
+
+            $classRecord->addImplements(FullyQualifiedName::fromString($interfaceName->getNamespacedName()->getFullyQualifiedNameText()));
+
+            $interfaceRecord = $this->index->get(ClassRecord::fromName($interfaceName));
+            assert($interfaceRecord instanceof ClassRecord);
+            $interfaceRecord->addImplementation($classRecord->fqn());
+
+            $this->index->write($interfaceRecord);
+        }
+    }
+
+    private function indexBaseClass(ClassRecord $record, ClassDeclaration $node): void
+    {
+        // @phpstan-ignore-next-line
+        if (null === $baseClause = $node->classBaseClause) {
+            return;
+        }
+
+        // @phpstan-ignore-next-line
```

Comment:
add comments about all ignores

Reply:
If it helps, my guess would be: because phpdoc does not contains `|null` so phpstan will complain about it ?

If yes, then I guess there is no need for further comments and given we can't change the phpdoc because we don't own the sources I guess we can't remove the comments neither ?
Maybe by adding some rules in the config file instead ?
Again, I'm not familiar with phpstan...

Reply:
Possibly - but I think it's more likely a problem with Phpstan generics. In this case it said the method should return a `Record` but it returns ` ClassRecord` - which is a subtype of Record.

As far as I know `@template TValue of Record` should allow the returning of `ClassRecord` (and it works in other places)

Reply:
Sorry, I was responding about a different ignore: _this_ one is because the Tolerarent Parser _says_ `baseClause` can never be NULL. But it's a lie :D and all the properties are nullable, so I'm just always defensive here.

Reply:
Added comments